albatross/tune.h

```diff
@@ -38,122 +34,127 @@ inline std::string to_string(const nlopt::result result) {
+  config.output_stream << model.pretty_string() << std::endl;
   config.output_stream << "objective: " << metric << std::endl;
   config.output_stream << "-------------------" << std::endl;
   return metric;
 }
 
-template <class FeatureType>
-ParameterStore
-tune_regression_model(const TuneModelConfig<FeatureType> &config) {
-
-  const auto example_model = config.model_creator();
-  auto x = example_model->get_tunable_parameters().values;
-
-  assert(x.size());
-  nlopt::opt opt = config.optimizer;
-  opt.set_min_objective(objective_function<FeatureType>, (void *)&config);
-  double minf;
-  nlopt::result result = opt.optimize(x, minf);
-
-  // Tell the user what the final parameters were.
-  example_model->set_tunable_params_values(x);
-  config.output_stream << "==================" << std::endl;
-  config.output_stream << "TUNED MODEL PARAMS" << std::endl;
-  config.output_stream << "nlopt termination code: " << to_string(result)
-                       << std::endl;
-  config.output_stream << "==================" << std::endl;
-  config.output_stream << example_model->pretty_string() << std::endl;
-
-  return example_model->get_params();
+template <typename ModelType, typename MetricType, class FeatureType>
+struct ModelTuner {
+  ModelType model;
+  MetricType metric;
+  std::vector<RegressionDataset<FeatureType>> datasets;
+  TuningMetricAggregator aggregator;
+  std::ostream &output_stream;
+  nlopt::opt optimizer;
+
+  ModelTuner(const ModelType &model_, const MetricType &metric_,
+             const std::vector<RegressionDataset<FeatureType>> &datasets_,
+             const TuningMetricAggregator &aggregator_,
+             std::ostream &output_stream_)
+      : model(model_), metric(metric_), datasets(datasets_),
+        aggregator(aggregator_), output_stream(output_stream_), optimizer() {
+    initialize_optimizer();
+  };
+
+  ParameterStore tune() {
+    auto x = model.get_tunable_parameters().values;
+
+    assert(x.size());
+    optimizer.set_min_objective(
+        objective_function<ModelType, MetricType, FeatureType>, (void *)this);
+    double minf;
+    nlopt::result result = optimizer.optimize(x, minf);
+
+    // Tell the user what the final parameters were.
+    model.set_tunable_params_values(x);
+    output_stream << "==================" << std::endl;
+    output_stream << "TUNED MODEL PARAMS" << std::endl;
+    output_stream << "nlopt termination code: " << to_string(result)
+                  << std::endl;
+    output_stream << "==================" << std::endl;
+    output_stream << model.pretty_string() << std::endl;
+
+    return model.get_params();
+  }
+
+  void
+  initialize_optimizer(const nlopt::algorithm &algorithm = nlopt::LN_PRAXIS) {
+    // The various algorithms in nlopt are coded by the first two characters.
+    // In this case LN stands for local, gradient free.
+    auto tunable_params = model.get_tunable_parameters();
+
+    optimizer = nlopt::opt(algorithm, (unsigned)tunable_params.values.size());
```

Comment:
Since the algorithm is part of the optimizer, it looks like we can set it without changing the default in Albatross, but it requires knowing the number of dimensions. I've been changing the hardcoded default to NelderMead.
Do you think it's worth either changing the default algorithm, or adding a helper function to change the algorithm of the optimizer without the caller specifying (perhaps wrongly) the number of tunable parameters?

Reply:
What I've been doing is:
```
auto tuner = get_tuner(model, metric, dataset);
tuner.initialize_optimizer(nlopt::LN_NEALDERMEAD);
```
which will take care of the parameter size.  We could rename `initialize_optimizer` to `set_algorithm`? Though that's actually a little misleading since it does do a lot of other non-algorithm specific initializations.

Reply:
It didn't dawn on me that I could call `initialize_optimizer` again. But I prefer that name to `set_algorithm`. Looks like my request is already solved. (Assuming I can remember the solution).